src/ristretto/bulletproofs_plus.rs

```diff
@@ -0,0 +1,629 @@
+        let statement = RangeStatement::init(self.generators.clone(), vec![commitment], vec![None], None)
+            .map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?;
+
+        let proof = RistrettoRangeProof::prove(self.transcript_label, &statement, &witness)
+            .map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?;
+
+        Ok(proof.to_bytes())
+    }
+
+    fn verify(&self, proof: &Self::Proof, commitment: &HomomorphicCommitment<Self::PK>) -> bool {
+        return match RistrettoRangeProof::from_bytes(proof)
+            .map_err(|e| RangeProofError::InvalidRangeProof(e.to_string()))
+        {
+            Ok(rp) => {
+                let statement = RangeStatement {
+                    generators: self.generators.clone(),
+                    commitments: vec![commitment.0.clone().into()],
+                    commitments_compressed: vec![*commitment.0.compressed()],
+                    minimum_value_promises: vec![None],
+                    seed_nonce: None,
+                };
+                match RistrettoRangeProof::verify_do_not_recover_masks(self.transcript_label, &[statement], &[
+                    rp.clone()
+                ]) {
+                    Ok(_) => true,
+                    Err(e) => {
+                        if self.generators.extension_degree() != rp.extension_degree() {
+                            error!(
+                                target: LOG_TARGET,
+                                "Generators' extension degree ({:?}) and proof's extension degree ({:?}) do not \
+                                 match; consider using a BulletproofsPlusService with a matching extension degree",
+                                self.generators.extension_degree(),
+                                rp.extension_degree()
+                            );
+                        }
+                        error!(target: LOG_TARGET, "Internal range proof error ({})", e.to_string());
+                        false
+                    },
+                }
+            },
+            Err(e) => {
+                error!(
+                    target: LOG_TARGET,
+                    "Range proof could not be deserialized ({})",
+                    e.to_string()
+                );
+                false
+            },
+        };
+    }
+
+    fn range(&self) -> usize {
+        self.generators.bit_length()
+    }
+}
+
+impl ExtendedRangeProofService for BulletproofsPlusService {
+    type K = RistrettoSecretKey;
+    type PK = RistrettoPublicKey;
+    type Proof = Vec<u8>;
+
+    fn construct_extended_proof(
+        &self,
+        masks: Vec<RistrettoExtendedMask>,
+        values: Vec<u64>,
+        min_value_promises: Vec<Option<u64>>,
+        seed_nonce: Option<Self::K>,
+    ) -> Result<Self::Proof, RangeProofError> {
+        if masks.is_empty() || values.is_empty() || masks.len() != values.len() {
```

Comment:
I'm sure you'll find it easier if you pass in one array of a struct that contains all 3.
```
struct ExtendedCommitmentPromise { masks: Mask, value: u64, minimum_value: u64 }
fn construct_extended_proof(commitments: Vec<ExtendedCommitmentPromise>, seed_nonce:..) 
```